### ISSUE TITLE
Printf format checks are skipped for constexpr format strings

### ISSUE THREAD

**Issue by paradust7** [llvm/llvm-project#55805]
demo.cpp:
```c++
#include <cstdio>

constexpr const char* foo() {
  return "%s %d";
}

struct Bar {
    static constexpr char value[] = {'%', 's', '%', 'd', '\0'};
};
constexpr char Bar::value[];

int main() {
   printf(foo(), "abc", "def");
   printf(Bar::value, "abc", "def");
   return 0;
}
```

Clang does not produce any warning. gcc 11.2.0 produces both warnings:
```
demo.cpp: In function ‘int main()’:
demo.cpp:13:14: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘const char*’ [-Wformat=]
demo.cpp:14:16: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘const char*’ [-Wformat=]
```

Clang version is:
```
clang version 15.0.0 (https://github.com/llvm/llvm-project.git a72cc958a386e5fc97e8c30137fb56eb77ef571c)
Target: x86_64-unknown-linux-gnu
Thread model: posix
```

**Comment by inclyc** [comment]
clang thinks these expressions are NOT literal, statically evaluated expressions are not considered as literal currently, and it fallbacks to warn you `-Wformat-nonliteral` (returns before "real" format string checking)

```C++
int main() {
   printf(foo(), "abc", "def"); // CXXMemberCallExprClass
   printf(Bar::value, "abc", "def"); // DeclRefExprClass
   return 0;
}
```

Related to https://github.com/llvm/llvm-project/issues/56583

See: https://godbolt.org/z/oKf6733K4

**Comment by inclyc** [comment]
May be helpful to see `clang/lib/Sema/SemaChecking.cpp`

```C++
// Determine if an expression is a string literal or constant string.
// If this function returns false on the arguments to a function expecting a
// format string, we will usually need to emit a warning.
// True string literals are then checked by CheckFormatString.
static StringLiteralCheckType
checkFormatStringExpr(Sema &S, const Expr *E, ArrayRef<const Expr *> Args,
                      Sema::FormatArgumentPassingKind APK, unsigned format_idx,
                      unsigned firstDataArg, Sema::FormatStringType Type,
                      Sema::VariadicCallType CallType, bool InFunctionCall,
                      llvm::SmallBitVector &CheckedVarArgs,
                      UncoveredArgHandler &UncoveredArg, llvm::APSInt Offset,
                      bool IgnoreStringsWithoutSpecifiers = false) {

/* statements */
}
```

**Comment by inclyc** [comment]
https://reviews.llvm.org/D130906
